src/ir_pattern.py

```diff
@@ -0,0 +1,465 @@
+                state.map_location(p) for p in pattern.replacement_instr.inputs
+            }
+            pattern_outputs = LocationRefSetDict()
+            for i, out in enumerate(reversed(pattern.replacement_instr.outputs)):
+                out_ref = InstrRef(
+                    pattern_node, len(pattern_node.block.instructions) - 1 - i
+                )
+                out_instr = out_ref.instruction()
+                assert out_instr.mnemonic == "out.fictive" and out_instr.inputs == [out]
+                pattern_outputs.extend(
+                    state.map_location(out),
+                    pattern.flow_graph.instr_inputs[out_ref].get(out),
+                )
+
+            refs_to_replace: List[InstrRef] = []
+            for i, pat_instr in reversed(
+                list(enumerate(pattern_node.block.instructions))
+            ):
+                if pat_instr.mnemonic in ("in.fictive", "out.fictive"):
+                    continue
+                pat_ref = InstrRef(pattern_node, i)
+                cand_ref = state.map_ref(pat_ref)
+                instr = cand_ref.instruction()
+
+                # Only add this instruction to refs_to_replace if its outputs are one of
+                # the replacement instruction's outputs, or if they are not used by by
+                # any instruction outside this pattern.
+                is_unreferenced = True
+                for reg, refs in flow_graph.instr_uses[cand_ref].items():
+                    if pat_ref in pattern_outputs.get(reg):
+                        continue
+                    if not all(r in refs_to_replace for r in refs):
+                        is_unreferenced = False
+                        break
+
+                if is_unreferenced:
+                    refs_to_replace.append(cand_ref)
+                elif any(r in pattern_inputs for r in instr.clobbers + instr.outputs):
+                    # If this instruction can't be replaced, but it clobbers a needed
+                    # input register, then we can't perform the rewrite
+                    # TODO: It may be possible to do the rewrite by introducing
+                    # additional temporary registers?
+                    refs_to_replace = []
+                    break
+            if not refs_to_replace:
+                continue
+
+            # Replace unreferenced instructions. The last instruction in the source is
+            for i, ref in enumerate(refs_to_replace):
+                # Remove ref from all instr_uses
+                instr = ref.instruction()
+                for _, rs in flow_graph.instr_inputs[ref].items():
+                    for r in rs:
+                        if isinstance(r, InstrRef):
+                            flow_graph.instr_uses[r].remove_ref(ref)
+
+                # The last instruction in the source is rewritten with the replacement instruction
+                # (refs_to_replace is in reverse order), and the others are replaced with a nop.
```

Comment:
Can it happen that replacing the last instruction results in a read from a totally different input? Say we have the pattern
```
mov $t, $i
mov $o, $t
->
mov $o, $i
```
and match that against the asm
```
mov $t, $i
li $i, 42
mov $o, $t
```
resulting in
```
nop
li $i, 42
mov $o, $i
```
Like in the comment above, new fictive temp registers or using the in/out graph in translate.py would help...